src/Orleans/Threading/CancellationTokenHolderGrain.cs

```diff
@@ -0,0 +1,50 @@
+﻿using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Orleans.Placement;
+
+namespace Orleans.Threading
+{
+    [PreferLocalPlacement]
+    internal class CancellationTokenHolderGrain : Grain, ICancellationTokenHolderGrain
+    {
+        private CancellationTokenSource cancellationTokenSource;
+        private bool collectionScheduled;
+
+        public override Task OnActivateAsync()
+        {
+            DelayDeactivation(TimeSpan.FromDays(5));
+            cancellationTokenSource = new CancellationTokenSource();
+            return TaskDone.Done;
+        }
+
+        public Task<CancellationToken> GetCancellationToken()
+        {
+            return Task.FromResult(cancellationTokenSource.Token);
+        }
+
+        public Task Cancel()
+        {
+            cancellationTokenSource.Cancel();
+            DelayDeactivation(TimeSpan.FromSeconds(30));
```

Comment:
Would using the configured timeout be a lot of work? Say something like
```
var safetyCautionLimit = TimeSpan.FromSeconds(1);
DelayDeactivation(configuredDelay.Add(safetyCautionLimit));
```
I don't know what would the consequence here could be if that hardcoded 30 seconds would be off, so I my gut reaction might be just that.